### ISSUE TITLE
Add Epsilon Technique to Loss function to avoid log(0)

### ISSUE THREAD

**Issue by jose-compu** [jo-tm/micro-llm#6]
When using BinaryCrossEntropyLoss, PyTorch uses a small epsilon value to clip the predictions, ensuring that neither zero nor one probabilities are used directly. This prevents taking the log of zero.
For example:
```python
epsilon = 1e-12
pred = torch.clamp(pred, epsilon, 1. - epsilon)
```
This approach ensures that probabilities are bounded away from zero, maintaining numerical stability.

More detailed example:

```python
import torch
import torch.nn.functional as F

# Predicted probabilities (before applying sigmoid)
logits = torch.tensor([-10.0, 0.0, 10.0])

# Applying sigmoid to get probabilities
probs = torch.sigmoid(logits)

# Adding a small epsilon to avoid log(0)
epsilon = 1e-12
probs = torch.clamp(probs, epsilon, 1. - epsilon)

# Target values
targets = torch.tensor([0.0, 1.0, 0.0])

# Compute Binary Cross-Entropy Loss
loss = F.binary_cross_entropy(probs, targets)
print(loss)
```
